[박민서] 7주차 과제 제

m2nsp opened the pull request:
<!--
제목 예시: [이름] n주차 과제 제출
-->

## 📌 구현 기능  
- 게시물 좋아요
![스크린샷 2025-05-18 022521](https://github.com/user-attachments/assets/fc6fcb17-894a-43d2-abc2-ffe2ff4a1651)

- 게시물 좋아요 취소
![스크린샷 2025-05-18 022931](https://github.com/user-attachments/assets/76cafa39-0d2c-4d28-9625-a680c39a173d)
좋아요

- 댓글 삭제
![스크린샷 2025-05-18 023418](https://github.com/user-attachments/assets/c8eac46c-458c-4379-a715-a3ded6041d55)

## 🗂️ 과제 정리  
- 
- 

## ⚠️ 어려웠던 점  
- ~~생략 가능~~
- 댓글 수정 부분에서 commentId를 받는데 기존의 commentRequestDTO를 사용하니까 pathVariable로 받은 commentId가 필요가 없어지거나 그 반대가 되는 것 같은데...그럼 어떻게 하는 게 좋은 건지 궁금합니다... 아니면 그냥 commentId만 살리고 commentRepository에서 findById로 하는게 더 좋은 방법일까요..?
- 좋아요 수나 댓글 수를 보여주고 싶을 때 세션에서는 dto에서 넘겨주는 방법으로 설명해주셨었는데 글 수가 많아 진다면 DB에 칼럼으로 저장해 두는 게 오히려 부하가 적어지지 않을까 생각했는데 이 부분에 대해서도 궁금합니다!

m2nsp committed:
✨feat: 7주차 과제

m2nsp committed:
Merge branch 'main' of https://github.com/m2nsp/efub5-backend-assignment

s2eojeong commented on src/main/java/efub/assignment/community/post/domain/PostLike.java:
```diff
@@ -0,0 +1,32 @@
+import lombok.NoArgsConstructor;
+
+@NoArgsConstructor
+@AllArgsConstructor
```
`@Builder`가 명시적 생성자를 이미 만들어주고 있기 때문에 `@AllArgsConstructor`는 사실상 쓸모가 없을 것 같습니다. 지워주세요!

s2eojeong commented on src/main/java/efub/assignment/community/post/domain/PostLike.java:
```diff
@@ -0,0 +1,32 @@
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+
+@NoArgsConstructor
```
JPA에서 권장되는 기본 생성자 선언은 불필요하게 외부에서 객체 생성되는 걸 막기 위해 `@NoArgsConstructor(access = AccessLevel.PROTECTED)` 으로 쓰는 것을 추천드립니다.
* 참고로 이를 명시하지 않으면 디폴트 값은 `public` 입니다.

s2eojeong commented on src/main/java/efub/assignment/community/post/domain/PostLike.java:
```diff
@@ -0,0 +1,32 @@
+    private Long likeId;
+
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name="post_like_id", updatable = false, nullable = false)
```
`PostLike`는 `Post` 엔티티에 `@ManyToOne`으로 연관되어 있기 떄문에 이는 Post의 PK(`post_id`)를 참조하는 외래 키입니다.
현재 데이터베이스 관점에서 이 외래 키는 `Post` 테이블의 PK인 `post_id`를 참조하기 떄문에 컬럼 이름을 `post_id`로 명시해주셔야 합니다!

s2eojeong commented on src/main/java/efub/assignment/community/post/domain/PostLike.java:
```diff
@@ -0,0 +1,32 @@
+@Getter
+@Entity
+public class PostLike extends BaseEntity {
+    @Id @GeneratedValue(strategy = GenerationType.IDENTITY)
```
이 필드가 밑에서 쓰신 `post_like_id`에 해당됩니다.
때문에 여기에 `@Column(name = "post_like_id")`를 추가해주시면 될 것 같습니다.

s2eojeong commented on src/main/java/efub/assignment/community/post/service/PostService.java:
```diff
@@ -95,10 +99,46 @@ public void deletePost(Long boardId, Long postId){
+    @Transactional
+    public void likePost(Long postId, Long memberId) {
+        Post post = postValidation(postId);
+        Member member = memberRepository.findById(memberId)
```
`MemberService`에 `memberId`로 사용자를 찾는 로직을 이미 구현하셨던데 그 함수를 재사용하는건 어떨까요?
`Member member = memberService.findMemberByMemberId(memberId);`

s2eojeong commented on src/main/java/efub/assignment/community/post/service/PostService.java:
```diff
@@ -95,10 +99,46 @@ public void deletePost(Long boardId, Long postId){
+    @Transactional
+    public void unlikePost(Long postId, Long memberId) {
+        Post post = postValidation(postId);
+        Member member = memberRepository.findById(memberId)
```
여기서도 `Member member = memberService.findMemberByMemberId(memberId);` 로 바꾸면 더 깔끔할 것 같습니다!

s2eojeong commented on src/main/java/efub/assignment/community/post/service/PostService.java:
```diff
@@ -19,11 +21,13 @@ public class PostService {
     private final PostRepository postRepository;
     private final BoardRepository boardRepository;
     private final MemberRepository memberRepository;
+    private final PostLikeRepository postLikeRepository;
```
그러기 위해서 여기에 `private final MemberService memberService;`를 추가해주시면 될 것 같습니다.

s2eojeong commented on src/main/java/efub/assignment/community/post/service/PostService.java:
```diff
@@ -95,10 +99,46 @@ public void deletePost(Long boardId, Long postId){
+        Member member = memberRepository.findById(memberId)
+                .orElseThrow(() -> new IllegalArgumentException("해당 사용자를 찾을 수 없습니다."));
+        //좋아요 존재 여부 확인
+        if (!postLikeRepository.existsByPostAndMember(post, member)) {
```
현재 코드는 동일한 조건으로 두 번 확인하고 있는걸로 보이는데,
```java
if (!postLikeRepository.existsByPostAndMember(post, member)) {
    throw new IllegalArgumentException("해당 좋아요는 존재하지 않습니다.");
}
```
이 부분은 삭제하셔도 무방할 것 같습니다. 아래 코드에서 이미 예외 처리를 하고 있어서요.

s2eojeong commented on src/main/java/efub/assignment/community/post/controller/PostController.java:
일반적으로 RESTful API에서는 URL 끝에 슬래시(/)를 붙이지 않는 것이 관례입니다.
`게시글 생성`과 `게시글 목록 조회`에서 각각 `@PostMapping`과 `@GetMapping`으로 바꿔주세요!

s2eojeong commented on src/main/java/efub/assignment/community/comment/domain/Comment.java:
맨 위에 `@Getter` 추가해주시고 아래 get 메소드들은 삭제하시면 더 깔끔할 것 같습니다.

s2eojeong commented on src/main/java/efub/assignment/community/comment/service/CommentService.java:
과제 요구사항을 확인했는데, 댓글 삭제 부분이 구현이 아직 안된 것 같아요. 확인 부탁드립니다!

s2eojeong commented on src/main/java/efub/assignment/community/comment/controller/CommentController.java:
과제 요구사항을 확인했는데 아직 댓글 삭제 부분이 구현이 안된 것 같습니다. 확인 부탁드려요!

s2eojeong commented on src/main/java/efub/assignment/community/comment/service/CommentService.java:
`deleteComment` 에서 해당 `comment`를 작성한 사용자(accountId)가 현재 로그인된 사용자(accountId)와 일치하는지도 체크해야할 것 같아요.
```java
public void deleteComment(Long commentId, Long accountId){
    Comment comment = findByCommentId(commentId);
    Account account = accountService.findByAccountId(accountId);
    Comment comment = commentRepository.findByCommentAndAccount(comment, account)
                    .orElseThrow(() -> new IllegalArgumentException("해당 댓글이 존재하지 않습니다."));
    commentRepository.delete(comment);
}
```
이런식으로 수정해주시면 좋을 것 같습니다.

그리고 commentVlidation 부분을 따로 뺀다고 하면(다른 부분에서도 해당 유효성 검증 부분이 여러번 쓰이기 때문에)
```java
public Comment commentValidation(Long commentId, Long accountId) {
    Comment comment = findByCommentId(commentId);
    Account account = accountService.findByAccountId(accountId);

    return commentRepository.findByCommentAndAccount(comment, account)
                .orElseThrow(() -> new IllegalArgumentException("해당 댓글이 존재하지 않습니다."));
}
```
라고 고쳐주시면 앞에 deleteComment 부분도
```java
public void deleteComment(Long commentId, Long accountId){
    Comment comment = commentValidation(commentId, accountId);
    commentRepository.delete(comment);
}
```
과 같이 더 깔끔하게 코드를 작성할 수 있습니다!

s2eojeong reviewed (approved):
과제하느라 수고 많으셨습니다 !
코멘트 남겼으니 혹시 다른 질문 있으시면 편하게 질문 주세요 🤓

+) 두번째 질문에 답변을 드리자면, 현재는 간단한 프로젝트를 구현하는 단계여서 데이터 양이 많지 않기 때문에 DTO에 댓글 수나 좋아요 수를 계산해서 담는 방식으로도 충분히 괜찮습니다. 하지만 민서님이 말씀하신 것처럼 실무에서는 더 방대한 양의 데이터를 다루기 때문에 DB에 컬럼을 따로 만들어두기도 합니다. 서비스 규모에 따라 적재적소로 데이터를 저장하고 가져온다고 보시면 될 것 같습니다 !

s2eojeong commented on src/main/java/efub/assignment/community/comment/service/CommentService.java:
```diff
@@ -70,7 +70,7 @@ public CommentListResponseDTO getMemberCommentList(Long memberId) {
 
     @Transactional
-    public CommentResponseDTO updateComment(Long postId, Long commentId, CommentRequestDTO requestDTO) {
+    public CommentResponseDTO updateComment(Long commentId, CommentRequestDTO requestDTO) {
```
Q. 댓글 수정 부분에서 commentId를 받는데 기존의 commentRequestDTO를 사용하니까 pathVariable로 받은 commentId가 필요가 없어지거나 그 반대가 되는 것 같은데...그럼 어떻게 하는 게 좋은 건지 궁금합니다... 아니면 그냥 commentId만 살리고 commentRepository에서 findById로 하는게 더 좋은 방법일까요..?

A. DTO에는 수정에 필요한 데이터만 넣는 게 깔끔하기 때문에 `commentUpdateRequestDto`를 생성하는게 어떨까요? 저희는 `content`만 수정하면 되기 때문에 해당 필드만 명시하시면 될 것 같아요 !

s2eojeong commented on src/main/java/efub/assignment/community/post/repository/PostLikeRepository.java:
```diff
@@ -0,0 +1,14 @@
+public interface PostLikeRepository extends JpaRepository<PostLike, Long> {
+    boolean existsByPostAndMember(Post post, Member member);
+    Optional<PostLike> findByPostAndMember(Post post, Member member);
+    void deleteAllByPost(Post post);
```
`PostLike`가 `Post`에 `@OneToMany(mappedBy = "post", cascade = CascadeType.ALL, orphanRemoval = true)`로 연관되어 있고 양방향 매핑이 되어 있다면 굳이 `deleteAllByPost`를 만들지 않아도 됩니다 !

m2nsp committed:
♻️refactor: week7 - 코드리뷰 기반 리팩토링

- 댓글 작성자 맞는 지 확인 로직 추가
- post-postlike 제약 조건 추가
- 끝에 '/' url 경로 수정
- 기본생성자 protected 조건 설정 추가
- member 존재 여부 확인 로직 memberService모듈 재사용으로 변경
